lib/internal/wrap_js_stream.js

```diff
@@ -0,0 +1,220 @@
+  }
+
+  // Legacy
+  static get StreamWrap() {
+    return JSStreamWrap;
+  }
+
+  isAlive() {
+    return true;
+  }
+
+  isClosing() {
+    return !this.readable || !this.writable;
+  }
+
+  readStart() {
+    this.stream.resume();
+    return 0;
+  }
+
+  readStop() {
+    this.stream.pause();
+    return 0;
+  }
+
+  doShutdown(req) {
+    const handle = this._handle;
+    const item = this._enqueue('shutdown', req);
+
+    this.stream.end(() => {
+      // Ensure that write was dispatched
+      setImmediate(() => {
+        if (!this._dequeue(item))
+          return;
+
+        handle.finishShutdown(req, 0);
+      });
+    });
+    return 0;
+  }
+
+  doWrite(req, bufs) {
+    const self = this;
+    const handle = this._handle;
+
+    var pending = bufs.length;
+
+    // Queue the request to be able to cancel it
+    const item = this._enqueue('write', req);
+
+    this.stream.cork();
+    for (var n = 0; n < bufs.length; n++)
+      this.stream.write(bufs[n], done);
+    this.stream.uncork();
+
+    function done(err) {
+      if (!err && --pending !== 0)
+        return;
+
+      // Ensure that this is called once in case of error
+      pending = 0;
+
+      let errCode = 0;
+      if (err) {
+        const code = uv[`UV_${err.code}`];
+        errCode = (err.code && code) ? code : uv.UV_EPIPE;
+      }
+
+      // Ensure that write was dispatched
+      setImmediate(function() {
+        // Do not invoke callback twice
+        if (!self._dequeue(item))
+          return;
+
+        handle.doAfterWrite(req);
+        handle.finishWrite(req, errCode);
+      });
+    }
+
+    return 0;
+  }
+
+  _enqueue(type, req) {
+    const item = new QueueItem(type, req);
+    if (this._list === null) {
+      this._list = item;
+      return item;
+    }
+
+    item.next = this._list.next;
+    item.prev = this._list;
+    item.next.prev = item;
+    item.prev.next = item;
+
+    return item;
+  }
+
+  _dequeue(item) {
+    assert(item instanceof QueueItem);
+
+    var next = item.next;
+    var prev = item.prev;
+
+    if (next === null && prev === null)
+      return false;
+
+    item.next = null;
+    item.prev = null;
+
+    if (next === item) {
+      prev = null;
+      next = null;
+    } else {
+      prev.next = next;
+      next.prev = prev;
+    }
+
+    if (this._list === item)
+      this._list = next;
+
+    return true;
+  }
+
+  doClose(cb) {
+    const handle = this._handle;
+
+    setImmediate(() => {
+      while (this._list !== null) {
+        const item = this._list;
+        const req = item.req;
+        this._dequeue(item);
+
+        const errCode = uv.UV_ECANCELED;
+        if (item.type === 'write') {
+          handle.doAfterWrite(req);
+          handle.finishWrite(req, errCode);
+        } else if (item.type === 'shutdown') {
+          handle.finishShutdown(req, errCode);
+        }
+      }
+
+      // Should be already set by net.js
+      assert.strictEqual(this._handle, null);
+      cb();
+    });
+  }
```

Comment:
this function should really be top-level. The while loop can really get cpu-intensive.